fraud-detection-python/cloud-functions/process-invoices/main.py

```diff
@@ -0,0 +1,305 @@
+import re
+import os
+import json
+from typing import List, Tuple
+
+from google.cloud import bigquery
+from google.cloud import documentai_v1 as documentai
+from google.cloud import storage
+from google.cloud import pubsub_v1
+from google.api_core.operation import Operation
+
+# Reading environment variables
+gcs_output_uri_prefix = os.environ.get('GCS_OUTPUT_URI_PREFIX')
+project_id = os.environ.get('GCP_PROJECT')
+location = os.environ.get('PARSER_LOCATION')
+processor_id = os.environ.get('PROCESSOR_ID')
+geocode_request_topicname = os.environ.get('GEOCODE_REQUEST_TOPICNAME')
+timeout = int(os.environ.get('TIMEOUT'))
+
+# An array of Future objects
+# Every call to publish() returns an instance of Future
+geocode_futures = []
+kg_futures = []
+# Setting variables
+address_fields = ['receiver_address', 'remit_to_address',
+                  'ship_from_address', 'ship_to_address', 'supplier_address']
+
+# GCS Variables
+gcs_output_bucket = f"{project_id}-output-invoices"
+gcs_archive_bucket_name = f"{project_id}-archived-invoices"
+destination_uri = f"gs://{gcs_output_bucket}/{gcs_output_uri_prefix}/"
+
+dataset_name = 'invoice_parser_results'
+entities_table_name = 'doc_ai_extracted_entities'
+
+client_options = {
+    "api_endpoint": f"{location}-documentai.googleapis.com"
+}
+
+docai_client = documentai.DocumentProcessorServiceClient(
+    client_options=client_options)
+
+storage_client = storage.Client()
+bq_client = bigquery.Client()
+pub_client = pubsub_v1.PublisherClient()
+
+ACCEPTED_MIME_TYPES = set(['application/pdf', 'image/jpeg',
+                           'image/png', 'image/tiff', 'image/gif'])
+
+
+def write_to_bq(dataset_name, table_name, entities_extracted_dict):
+    """
+    Write Data to BigQuery
+    """
+    dataset_ref = bq_client.dataset(dataset_name)
+    table_ref = dataset_ref.table(table_name)
+    row_to_insert = []
+    row_to_insert.append(entities_extracted_dict)
+
+    json_data = json.dumps(row_to_insert, sort_keys=False)
+    # Convert to a JSON Object
+    json_object = json.loads(json_data)
+
+    schema_update_options = [bigquery.SchemaUpdateOption.ALLOW_FIELD_ADDITION,
+                             bigquery.SchemaUpdateOption.ALLOW_FIELD_RELAXATION]
+    source_format = bigquery.SourceFormat.NEWLINE_DELIMITED_JSON
+
+    job_config = bigquery.LoadJobConfig(
+        schema_update_options=schema_update_options,
+        source_format=source_format
+    )
+
+    job = bq_client.load_table_from_json(
+        json_object, table_ref, job_config=job_config)
+    return job.result()  # Waits for table load to complete.
+
+
+def extract_document_entities(document: documentai.Document) -> dict:
+    """
+    Get all entities from a document and output as a dictionary
+    Flattens nested entities/properties
+    Format: entity.type_: entity.mention_text OR entity.normalized_value.text
+    """
+    document_entities = {}
+
+    def extract_document_entity(entity: documentai.Document.Entity):
```

Comment:
Nit: having a helper function inside of a function is too congested for my opinion maybe move out this function.